xdsl/dialects/memref.py

```diff
@@ -587,6 +587,32 @@ def get(
         return Cast.build(operands=[source], result_types=[type])
 
 
+@irdl_op_definition
+class MemorySpaceCast(IRDLOperation):
+    name = "memref.memory_space_cast"
+
+    source = operand_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+    dest = result_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+
+    def __init__(
+        self,
+        source: SSAValue | Operation,
+        dest_space: Attribute,
+    ):
+        source = SSAValue.get(source)
+
+        assert isinstance(source.type, MemRefType)
```

Comment:
this is not true in the general case, as during rewrites the types might not be the expected ones